### ISSUE TITLE
Entropy scoring overhaul: semantic emotion detection via Plutchik + local embeddings

### ISSUE THREAD

**Issue by coolmanns** [CoderofTheWest/openclaw-plugin-stability#4]
## Hey Chris 👋

Sascha here (coolmanns fork). We spent a solid evening deep-diving into the entropy scoring system and ended up rebuilding the emotion detection layer. Wanted to share what we found and what we built — might be useful for Clint too.

## The Problem

The keyword-based emotion scoring was missing real user frustration. Messages like "I'm super frustrated" and "this is sooo unacceptable" scored **0.00** because the emotion keyword list was tuned for the Oct 31 Strange Loop event — almost entirely positive/observer-perspective words ("proud of you", "impressed", "breakthrough"). The downstream machinery (growth vectors, contemplation thresholds, metabolism gates) was starved by the upstream signal.

We also found that all categories were boolean — a single keyword hit gave the full category score. "frustrated broken stuck annoying" scored the same as just "stuck."

### 1. Upstream alignment
- Correction weight: +0.40 → +0.26 (matches Clint)
- `calculateShannonEntropy()` activated (was dead code) — provides novelty boost or repetition penalty
- Added entropy debt tracking from your paper: cumulative ledger, 0.05/turn bleed-off, 3-turn rolling avg >0.2 triggers heightened sensitivity

### 2. Intensity scaling
All categories switched from boolean to count-based:
```js
Math.min(baseScore + hitCount * increment, cap)
```
Emotions: base 0.15, +0.10/hit, cap 0.50. Corrections: base 0.15, +0.07/hit, cap 0.40.

### 3. Semantic emotion scoring (the big one)
Replaced the keyword emotion block with **Plutchik emotion wheel + local embeddings**:

- 8 primary emotions × 3 intensity levels = 24 anchors
- Each anchor has 4-8 real conversational phrases (109 total)
- Pre-computed 768-dim embeddings via nomic-embed (local, ~650KB JSON file)
- Runtime: one HTTP call to embedding model (~5ms), cosine similarity against all phrase embeddings
- **No LLM calls, no tokens, no cost** — pure embedding math

This catches what keywords never could:
| Message | Keyword score | Semantic score |
|---------|:---:|:---:|
| "I'm super frustrated" | 0.00 | 0.325 |
| "This is fucking amazing!" | 0.00 | 0.471 |
| "This is sooo unacceptable" | 0.00 | 0.515 |
| "useless waste of time" | 0.00 | 0.220 |
| "holy shit that actually worked" | 0.00 | 0.290 |
| "yes please" | 0.00 | 0.045 |
| "Approved. Go ahead." | 0.00 | 0.011 |

### 4. Valence gate
Embeddings measure arousal, not direction. "Oh my god this is frustrating" matched joy anchors because the arousal profile is similar. Added a pure JS valence gate — curated positive/negative word lists determine message polarity, then reroute anchor selection when the embedding's best match has conflicting valence.

Negative families: anger, sadness, fear, disgust. Positive: joy, trust, anticipation.

Result: "frustrating make me cringe" went from joy_medium (0.098) → anger_medium (0.317).

### 5. Other fixes
- Stem-based regex matching for keyword fallback path
- Ring buffer expanded 5 → 50 entries
- `metaConceptWarningThreshold` 10 → 3
- OpenClaw channel envelope stripping (Telegram metadata was diluting embeddings)

## Architecture

```
User message
  ↓
Strip channel envelope (Telegram metadata, audio headers)
  ↓
Embed via nomic-embed (768-dim, ~5ms local)
  ↓
Cosine similarity against 109 pre-computed phrase embeddings
  ↓
Best match → Plutchik anchor (emotion + intensity)
  ↓
Valence gate: does anchor polarity match message polarity?
  ├─ Yes → use score
  └─ No → reroute to best anchor of correct valence family
  ↓
Weighted score based on anchor intensity level
  ↓
Combined with Shannon entropy, correction detection, debt tracking
  ↓
Final composite entropy score
```

## Known Limitations

- **Sarcasm** — embeddings can't catch it. "So I need to curb my sarcasm" scores low.
- **Indirect frustration** — "this better be not another bad experience" → 0.049 (negation patterns are hard for embeddings)
- **Embedding dependency** — requires a local embedding model. We use nomic-embed via llama.cpp on port 8082. Not everyone has that.

## Your Paper Called It

Section 7.5 of your growth-vector-systems-paper notes that model-internal entropy (logit distributions) would be better than text-based pattern matching. We can't get at logits through the API, but local embeddings are a step in that direction — semantic similarity instead of string matching, staying within the text-based paradigm.

## Fork

Our fork: [coolmanns/openclaw-plugin-stability](https://github.com/coolmanns/openclaw-plugin-stability)

Not submitting as a PR because the changes are opinionated (nomic-embed dependency, channel-specific stripping), but happy to collaborate if you want to pull any of this upstream. The Plutchik anchor approach + valence gate could work with any embedding backend.

— Sascha & Gandalf 🧙‍♂️
